From ec14a305bd3d4579655aa9f2d80b904c0e514390 Mon Sep 17 00:00:00 2001
From: Charles Keepax <ckeepax@opensource.wolfsonmicro.com>
Date: Tue, 25 Feb 2014 14:21:30 +0000
Subject: [PATCH 136/351] switch: arizona: Allow headphone detection to be
 skipped

Add a pdata/DT option to allow headphone detection to be skipped
entirely.

Change-Id: I0db05f532f5023fef1c80c79fa08a598814924b7
Signed-off-by: Charles Keepax <ckeepax@opensource.wolfsonmicro.com>
Signed-off-by: Nikesh Oswal <Nikesh.Oswal@wolfsonmicro.com>
---
 Documentation/devicetree/bindings/mfd/arizona.txt |    4 ++++
 drivers/switch/switch-arizona.c                   |   24 ++++++++++++++++++---
 include/linux/mfd/arizona/pdata.h                 |    3 +++
 3 files changed, 28 insertions(+), 3 deletions(-)

diff --git a/Documentation/devicetree/bindings/mfd/arizona.txt b/Documentation/devicetree/bindings/mfd/arizona.txt
index 8a25210..fd0c45e 100644
--- a/Documentation/devicetree/bindings/mfd/arizona.txt
+++ b/Documentation/devicetree/bindings/mfd/arizona.txt
@@ -109,6 +109,9 @@ Optional properties:
 
     wlf,wm5102t-output-pwr : Output power setting (WM5102T only)
 
+  - wlf,fixed-hpdet-imp : Do not perform any headphone detection, just use
+    the fixed value specified here as the headphone impedance.
+
 Example:
 
 codec: wm5102@1a {
@@ -139,6 +142,7 @@ codec: wm5102@1a {
 		0x1 1 0
 		0x0 2 1
 	>;
+	wlf,fixed-hpdet-imp = <8>;
 
 	wlf,micbias2 = <2600 0 1 1 0>;
 	wlf,init-mic-delay = <10>;
diff --git a/drivers/switch/switch-arizona.c b/drivers/switch/switch-arizona.c
index 7ef5b0f..f04683b 100644
--- a/drivers/switch/switch-arizona.c
+++ b/drivers/switch/switch-arizona.c
@@ -796,6 +796,21 @@ static void arizona_identify_headphone(struct arizona_extcon_info *info)
 	if (info->hpdet_done)
 		return;
 
+	if (info->arizona->pdata.fixed_hpdet_imp) {
+		int imp = info->arizona->pdata.fixed_hpdet_imp;
+
+		switch (arizona->type) {
+		case WM5110:
+			arizona_wm5110_tune_headphone(info, imp);
+			info->arizona->hp_impedance = imp;
+			break;
+		default:
+			break;
+		}
+
+		goto out;
+	}
+
 	dev_dbg(arizona->dev, "Starting HPDET\n");
 
 	/* Make sure we keep the device enabled during the measurement */
@@ -814,7 +829,7 @@ static void arizona_identify_headphone(struct arizona_extcon_info *info)
 				 ARIZONA_ACCDET_MODE_HPL);
 	if (ret != 0) {
 		dev_err(arizona->dev, "Failed to set HPDETL mode: %d\n", ret);
-		goto err;
+		goto out;
 	}
 
 	ret = regmap_update_bits(arizona->regmap, ARIZONA_HEADPHONE_DETECT_1,
@@ -822,12 +837,12 @@ static void arizona_identify_headphone(struct arizona_extcon_info *info)
 	if (ret != 0) {
 		dev_err(arizona->dev, "Can't start HPDETL measurement: %d\n",
 			ret);
-		goto err;
+		goto out;
 	}
 
 	return;
 
-err:
+out:
 	regmap_update_bits(arizona->regmap, ARIZONA_ACCESSORY_DETECT_MODE_1,
 			   ARIZONA_ACCDET_MODE_MASK, ARIZONA_ACCDET_MODE_MIC);
 
@@ -1340,6 +1355,9 @@ static int arizona_extcon_get_pdata(struct arizona *arizona)
 	arizona_of_read_u32(arizona, "wlf,init-mic-delay", false,
 			    &pdata->init_mic_delay);
 
+	arizona_of_read_u32(arizona, "wlf,fixed-hpdet-imp", false,
+			    &pdata->fixed_hpdet_imp);
+
 	return 0;
 }
 
diff --git a/include/linux/mfd/arizona/pdata.h b/include/linux/mfd/arizona/pdata.h
index 8ac525f..ba174aaf 100644
--- a/include/linux/mfd/arizona/pdata.h
+++ b/include/linux/mfd/arizona/pdata.h
@@ -132,6 +132,9 @@ struct arizona_pdata {
 	/** Internal pull on GPIO5 is disabled when used for jack detection */
 	bool jd_gpio5_nopull;
 
+	/** If non-zero don't run headphone detection, report this value */
+	int fixed_hpdet_imp;
+
 	/** Use the headphone detect circuit to identify the accessory */
 	bool hpdet_acc_id;
 
-- 
1.7.9.5

